# __Analisis de Evasion de Clientes__

## Librerias

In [1]:
import pandas as pd

## Extraccion de datos

### Importar Datos

In [2]:
url = 'https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json'

df = pd.read_json(url)
df.head()


customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

Como vemos, hay varias variables que contienen diccionarios en sus valores. Por tanto, debemos expandirlas

In [3]:
# obtener dataframe de variables con diccionario
cols_dict = [col for col in df.columns if isinstance(df[col].iloc[0], dict)]
df_dict = df[cols_dict]
df_dict.head()

customer  \
0  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [4]:
# obtener dataframe con variables normales
df_var = df.drop(columns=cols_dict)
df_var.head()

customerID Churn
0  0002-ORFBO    No
1  0003-MKNFE    No
2  0004-TLHLJ   Yes
3  0011-IGKFF   Yes
4  0013-EXCHZ   Yes

In [5]:

from pandas import json_normalize

# normalizamos variables con diccionarios
df_norm = [json_normalize(df_dict[col]) for col in df_dict.columns]

# combinar variables normalizadas horizontalmente
df_norm = pd.concat(df_norm, axis=1)
df_norm.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes        Yes       9          Yes   
1    Male              0      No         No       9          Yes   
2    Male              0      No         No       4          Yes   
3    Male              1     Yes         No      13          Yes   
4  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   
3            No     Fiber optic             No          Yes              Yes   
4            No     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      PaymentMethod  Charges.Monthly Charges.Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85  
3  Electronic check             98.0       1237.85  
4      Mailed check             83.9         267.4

In [6]:
# unir ambos dataframes
df_exp = pd.concat([df_norm, df_var], axis=1)
df_exp.head()


gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes        Yes       9          Yes   
1    Male              0      No         No       9          Yes   
2    Male              0      No         No       4          Yes   
3    Male              1     Yes         No      13          Yes   
4  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup  ... TechSupport  \
0            No             DSL             No          Yes  ...         Yes   
1           Yes             DSL             No           No  ...          No   
2            No     Fiber optic             No           No  ...          No   
3            No     Fiber optic             No          Yes  ...          No   
4            No     Fiber optic             No           No  ...         Yes   

  StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes              No        One year              Yes   
1          No             Yes  Month-to-month               No   
2          No              No  Month-to-month              Yes   
3         Yes             Yes  Month-to-month              Yes   
4         Yes              No  Month-to-month              Yes   

      PaymentMethod Charges.Monthly  Charges.Total  customerID Churn  
0      Mailed check            65.6          593.3  0002-ORFBO    No  
1      Mailed check            59.9          542.4  0003-MKNFE    No  
2  Electronic check            73.9         280.85  0004-TLHLJ   Yes  
3  Electronic check            98.0        1237.85  0011-IGKFF   Yes  
4      Mailed check            83.9          267.4  0013-EXCHZ   Yes  

[5 rows x 21 columns]

### Contexto de los datos

Se tratan de datos de los clientes de una empresa cuya informacion de cada variable se interpreta como:

- _customerID_: número de identificación único de cada cliente
- _Churn_: si el cliente dejó o no la empresa
- _gender_: género (masculino y femenino)
- _SeniorCitizen_: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
- _Partner_: si el cliente tiene o no una pareja
- _Dependents_: si el cliente tiene o no dependientes
- _tenure_: meses de contrato del cliente
- _PhoneService_: suscripción al servicio telefónico
- _MultipleLines_: suscripción a más de una línea telefónica
- _InternetService_: suscripción a un proveedor de internet
- _OnlineSecurity_: suscripción adicional de seguridad en línea
- _OnlineBackup_: suscripción adicional de respaldo en línea
- _DeviceProtection_: suscripción adicional de protección del dispositivo
- _TechSupport_: suscripción adicional de soporte técnico, menor tiempo de espera
- _StreamingTV_: suscripción de televisión por cable
- _StreamingMovies_: suscripción de streaming de películas
- _Contract_: tipo de contrato
- _PaperlessBilling_: si el cliente prefiere recibir la factura en línea
- _PaymentMethod_: forma de pago
- _Charges.Monthly_: total de todos los servicios del cliente por mes
- _Charges.Total_: total gastado por el cliente

### Tipo de datos

In [7]:
df_exp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7267 non-null   object 
 1   SeniorCitizen     7267 non-null   int64  
 2   Partner           7267 non-null   object 
 3   Dependents        7267 non-null   object 
 4   tenure            7267 non-null   int64  
 5   PhoneService      7267 non-null   object 
 6   MultipleLines     7267 non-null   object 
 7   InternetService   7267 non-null   object 
 8   OnlineSecurity    7267 non-null   object 
 9   OnlineBackup      7267 non-null   object 
 10  DeviceProtection  7267 non-null   object 
 11  TechSupport       7267 non-null   object 
 12  StreamingTV       7267 non-null   object 
 13  StreamingMovies   7267 non-null   object 
 14  Contract          7267 non-null   object 
 15  PaperlessBilling  7267 non-null   object 
 16  PaymentMethod     7267 non-null   object 


> El conjunto de datos esta conformado por 7267 registros y 21 variables.

> Todas las variables tienes 7267 registros lo que indica la ausencia de valores faltantes.

> Las variables 'SeniorCitizen' y 'tenure' son de tipo entero, la variable 'Charges.Monthly' es de tipo float y el resto de variables son de tipo object.

> La variable 'Charges.Total' en la practica es de tipo numerico. Sin embargo, Pandas la detecta como de tipo object lo que sugiere un problema con la consistencia de los datos.

## Consistencia de los datos

### Registros duplicados

In [8]:
# verificar registros duplicados
df_exp.duplicated().sum()

0

> No hay registros duplicados en el dataframe.

### Tratar valores vacios

In [9]:
# verificar consistencia en variables de tipo object
print(df_exp.gender.unique())
print(df_exp.SeniorCitizen.unique())
print(df_exp.Partner.unique())
print(df_exp.Dependents.unique())
print(df_exp.PhoneService.unique())
print(df_exp.MultipleLines.unique())
print(df_exp.InternetService.unique())
print(df_exp.OnlineSecurity.unique())
print(df_exp.OnlineBackup.unique())
print(df_exp.DeviceProtection.unique())
print(df_exp.TechSupport.unique())
print(df_exp.StreamingTV.unique())
print(df_exp.StreamingMovies.unique())
print(df_exp.Contract.unique())
print(df_exp.PaperlessBilling.unique())
print(df_exp.PaymentMethod.unique())
print(df_exp['Charges.Total'].unique())
print(df_exp.Churn.unique())

['Female' 'Male']
[0 1]
['Yes' 'No']
['Yes' 'No']
['Yes' 'No']
['No' 'Yes' 'No phone service']
['DSL' 'Fiber optic' 'No']
['No' 'Yes' 'No internet service']
['Yes' 'No' 'No internet service']
['No' 'Yes' 'No internet service']
['Yes' 'No' 'No internet service']
['Yes' 'No' 'No internet service']
['No' 'Yes' 'No internet service']
['One year' 'Month-to-month' 'Two year']
['Yes' 'No']
['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)']
['593.3' '542.4' '280.85' ... '742.9' '4627.65' '3707.6']
['No' 'Yes' '']


> Los valores de _Churn_ presentan valores vacios.

> Los valores de _Charges.Total_ presentan valores vacios

 Obtengamos la cantidad de registros sin valor para ambas variables. 

In [10]:
# registros con valores vacios en _Churn_
df_exp[df_exp['Churn'] == '']

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
30    Female              0      No         No      11          Yes   
75      Male              0      No         No      71          Yes   
96      Male              0      No         No      29          Yes   
98    Female              1      No         No       5          Yes   
175     Male              1     Yes         No      65          Yes   
...      ...            ...     ...        ...     ...          ...   
7158  Female              0      No         No      14          Yes   
7180  Female              0     Yes         No      49           No   
7211    Male              0      No         No       9          Yes   
7239  Female              0     Yes        Yes      67          Yes   
7247  Female              0     Yes        Yes      31          Yes   

         MultipleLines InternetService       OnlineSecurity  \
30                 Yes     Fiber optic                  Yes   
75                  No              No  No internet service   
96                  No             DSL                  Yes   
98                  No             DSL                  Yes   
175                Yes     Fiber optic                   No   
...                ...             ...                  ...   
7158               Yes             DSL                   No   
7180  No phone service             DSL                  Yes   
7211               Yes     Fiber optic                   No   
7239                No              No  No internet service   
7247                No              No  No internet service   

             OnlineBackup  ...          TechSupport          StreamingTV  \
30                     No  ...                   No                   No   
75    No internet service  ...  No internet service  No internet service   
96                    Yes  ...                  Yes                   No   
98                    Yes  ...                  Yes                   No   
175                   Yes  ...                   No                  Yes   
...                   ...  ...                  ...                  ...   
7158                  Yes  ...                   No                   No   
7180                   No  ...                   No                  Yes   
7211                   No  ...                   No                   No   
7239  No internet service  ...  No internet service  No internet service   
7247  No internet service  ...  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
30                     No  Month-to-month              Yes   
75    No internet service        Two year              Yes   
96                     No        One year              Yes   
98                     No  Month-to-month               No   
175                   Yes        One year              Yes   
...                   ...             ...              ...   
7158                   No        One year              Yes   
7180                   No  Month-to-month               No   
7211                   No  Month-to-month              Yes   
7239  No internet service        Two year              Yes   
7247  No internet service  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  customerID  \
30    Bank transfer (automatic)           79.00          929.3  0047-ZHDTW   
75      Credit card (automatic)           19.90         1355.1  0120-YZLQA   
96             Electronic check           58.75         1696.2  0154-QYHJU   
98      Credit card (automatic)           59.90         287.85  0162-RZGMZ   
175   Bank transfer (automatic)          103.15        6792.45  0274-VVQOQ   
...                         ...             ...            ...         ...   
7158               Mailed check           54.25          773.2  9840-GSRFX   
7180  Bank transfer (automatic)           40.65        2070.75  9872-RZQQB   
7211           Electronic check     

> Hay 224 registros con valores vacios en _Churn_ lo que representa el 3% de los registros totales. Como la variable objetivo es _Churn_, sin esta informacion tales registros no aportan al analisis por tanto los eliminamos.

In [11]:
# elimina registros con Churn vacios
df_exp = df_exp[df_exp['Churn'] != '']
df_exp.shape

(7043, 21)

In [12]:
# verificar que Churn este libre de valores vacios
print(df_exp.Churn.unique())

['No' 'Yes']


In [13]:
# registros vacios en Charges.Total
df_exp[df_exp['Charges.Total'] == ' ']

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
975   Female              0     Yes        Yes       0           No   
1775  Female              0     Yes        Yes       0          Yes   
1955    Male              0      No        Yes       0          Yes   
2075    Male              0     Yes        Yes       0          Yes   
2232    Male              0      No        Yes       0          Yes   
2308    Male              0     Yes        Yes       0          Yes   
2930  Female              0     Yes        Yes       0          Yes   
3134    Male              0     Yes        Yes       0          Yes   
3203  Female              0     Yes        Yes       0           No   
4169  Female              0     Yes        Yes       0          Yes   
5599    Male              0     Yes        Yes       0          Yes   

         MultipleLines InternetService       OnlineSecurity  \
975   No phone service             DSL                  Yes   
1775                No              No  No internet service   
1955               Yes             DSL                  Yes   
2075                No              No  No internet service   
2232                No              No  No internet service   
2308               Yes              No  No internet service   
2930               Yes             DSL                   No   
3134               Yes              No  No internet service   
3203  No phone service             DSL                  Yes   
4169                No             DSL                  Yes   
5599                No              No  No internet service   

             OnlineBackup  ...          TechSupport          StreamingTV  \
975                   Yes  ...                  Yes                  Yes   
1775  No internet service  ...  No internet service  No internet service   
1955                  Yes  ...                  Yes                   No   
2075  No internet service  ...  No internet service  No internet service   
2232  No internet service  ...  No internet service  No internet service   
2308  No internet service  ...  No internet service  No internet service   
2930                  Yes  ...                  Yes                  Yes   
3134  No internet service  ...  No internet service  No internet service   
3203                   No  ...                  Yes                  Yes   
4169                  Yes  ...                   No                  Yes   
5599  No internet service  ...  No internet service  No internet service   

          StreamingMovies  Contract PaperlessBilling  \
975                    No  Two year               No   
1775  No internet service  Two year               No   
1955                   No  Two year              Yes   
2075  No internet service  One year              Yes   
2232  No internet service  Two year               No   
2308  No internet service  Two year               No   
2930                   No  Two year               No   
3134  No internet service  Two year               No   
3203                   No  Two year              Yes   
4169                  Yes  Two year               No   
5599  No internet service  Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  customerID  \
975     Credit card (automatic)           56.05                 1371-DWPAZ   
1775               Mailed check           20.00                 2520-SGTTA   
1955  Bank transfer (automatic)           61.90                 2775-SEFEE   
2075               Mailed check           19.70                 2923-ARZLG   
2232               Mailed check           20.25                 3115-CZMZD   
2308               Mailed check           25.35                 3213-VVOLG   
2930               Mailed check           73.35                 4075-WKNIU   
3134               Mailed check           25.75                 4367-NUYAO   
3203  Bank transfer (automatic)           52.55                 4472-LVYGI   
4169               Mailed chec

> Hay 11 registros con valores vacios en Charges.Total lo que representa el 0.1% de los registros totales. Debido a que son tan pocos y que se trata de una variable relevante,procedemos a eliminarlos. 

In [14]:
# elimina registros con Charges.Total vacios
df_exp = df_exp[df_exp['Charges.Total'] != ' ']
df_exp.shape

(7032, 21)

In [15]:
# verificar que Churn este libre de valores vacios
print(df_exp['Charges.Total'].unique())

['593.3' '542.4' '280.85' ... '742.9' '4627.65' '3707.6']


In [18]:
import numpy as np 

# convertir a tipo entero
df_exp['Charges.Total'] = df_exp['Charges.Total'].astype(np.float64)

# verificar tipo
df_exp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7032 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7032 non-null   object 
 1   SeniorCitizen     7032 non-null   int64  
 2   Partner           7032 non-null   object 
 3   Dependents        7032 non-null   object 
 4   tenure            7032 non-null   int64  
 5   PhoneService      7032 non-null   object 
 6   MultipleLines     7032 non-null   object 
 7   InternetService   7032 non-null   object 
 8   OnlineSecurity    7032 non-null   object 
 9   OnlineBackup      7032 non-null   object 
 10  DeviceProtection  7032 non-null   object 
 11  TechSupport       7032 non-null   object 
 12  StreamingTV       7032 non-null   object 
 13  StreamingMovies   7032 non-null   object 
 14  Contract          7032 non-null   object 
 15  PaperlessBilling  7032 non-null   object 
 16  PaymentMethod     7032 non-null   object 


### Columna de cuentas diarias

In [ ]:
df_exp

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes        Yes       9          Yes   
1       Male              0      No         No       9          Yes   
2       Male              0      No         No       4          Yes   
3       Male              1     Yes         No      13          Yes   
4     Female              1     Yes         No       3          Yes   
...      ...            ...     ...        ...     ...          ...   
7262  Female              0      No         No      13          Yes   
7263    Male              0     Yes         No      22          Yes   
7264    Male              0      No         No       2          Yes   
7265    Male              0     Yes        Yes      67          Yes   
7266    Male              0     Yes        Yes      63           No   

         MultipleLines InternetService OnlineSecurity OnlineBackup  ...  \
0                   No             DSL             No          Yes  ...   
1                  Yes             DSL             No           No  ...   
2                   No     Fiber optic             No           No  ...   
3                   No     Fiber optic             No          Yes  ...   
4                   No     Fiber optic             No           No  ...   
...                ...             ...            ...          ...  ...   
7262                No             DSL            Yes           No  ...   
7263               Yes     Fiber optic             No           No  ...   
7264                No             DSL             No          Yes  ...   
7265                No             DSL            Yes           No  ...   
7266  No phone service             DSL            Yes          Yes  ...   

     TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0            Yes         Yes              No        One year              Yes   
1             No          No             Yes  Month-to-month               No   
2             No          No              No  Month-to-month              Yes   
3             No         Yes             Yes  Month-to-month              Yes   
4            Yes         Yes              No  Month-to-month              Yes   
...          ...         ...             ...             ...              ...   
7262         Yes          No              No        One year               No   
7263          No          No             Yes  Month-to-month              Yes   
7264          No          No              No  Month-to-month              Yes   
7265         Yes          No             Yes        Two year               No   
7266          No         Yes             Yes        Two year               No   

         PaymentMethod Charges.Monthly  Charges.Total  customerID Churn  
0         Mailed check           65.60          593.3  0002-ORFBO    No  
1         Mailed check           59.90          542.4  0003-MKNFE    No  
2     Electronic check           73.90         280.85  0004-TLHLJ   Yes  
3     Electronic check           98.00        1237.85  0011-IGKFF   Yes  
4         Mailed check           83.90          267.4  0013-EXCHZ   Yes  
...                ...             ...            ...         ...   ...  
7262      Mailed check           55.15          742.9  9987-LUTYD    No  
7263  Electronic check           85.10         1873.7  9992-RRAMN   Yes  
7264      Mailed check           50.30          92.75  9992-UJOEL    No  
7265      Mailed check           67.85        4627.65  9993-LHIEB    No  
7266  Electronic check           59.00         3707.6  9995-HOTOH    No  

[7032 rows x 21 columns]

In [ ]:
# crear columna de gasto diario de cada cliente
df_exp['Cuentas_Diarias'] = (df_exp['Charges.Monthly'])/(df_exp['tenure']*30)
df_exp.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes        Yes       9          Yes   
1    Male              0      No         No       9          Yes   
2    Male              0      No         No       4          Yes   
3    Male              1     Yes         No      13          Yes   
4  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup  ... StreamingTV  \
0            No             DSL             No          Yes  ...         Yes   
1           Yes             DSL             No           No  ...          No   
2            No     Fiber optic             No           No  ...          No   
3            No     Fiber optic             No          Yes  ...         Yes   
4            No     Fiber optic             No           No  ...         Yes   

  StreamingMovies        Contract PaperlessBilling     PaymentMethod  \
0              No        One year              Yes      Mailed check   
1             Yes  Month-to-month               No      Mailed check   
2              No  Month-to-month              Yes  Electronic check   
3             Yes  Month-to-month              Yes  Electronic check   
4              No  Month-to-month              Yes      Mailed check   

  Charges.Monthly Charges.Total  customerID  Churn Cuentas_Diarias  
0            65.6        593.30  0002-ORFBO     No        0.242963  
1            59.9        542.40  0003-MKNFE     No        0.221852  
2            73.9        280.85  0004-TLHLJ    Yes        0.615833  
3            98.0       1237.85  0011-IGKFF    Yes        0.251282  
4            83.9        267.40  0013-EXCHZ    Yes        0.932222  

[5 rows x 22 columns]